Advanced features

Katze719 opened the pull request.

Katze719 committed:
add more functionaltity

Katze719 committed:
modernize

Katze719 committed:
refactor project name to lowercase and update references in CMake, README, CI, and examples

Katze719 committed:
refactor serial API to improve callback handling and update versioning structure

Katze719 committed:
enhance serial API to provide detailed port information via callback, updating example and tests accordingly

Katze719 committed:
remove unnecessary output for available serial ports in unit test

Katze719 committed:
implement serialGetPortsInfo in a new file, refactoring the logic for collecting USB device information and updating the callback mechanism; remove redundant implementation from serial.cpp

Katze719 committed:
update version to 1.0.0 in CMakeLists.txt and remove unused includes from serial.cpp

Katze719 committed:
Add comprehensive documentation for C API, building instructions, Deno examples, and usage guidelines

Katze719 committed:
Enhance CMake configuration for improved portability and CI workflow; enable position-independent code, link libraries statically, and update build steps in CI to use a container image.

Katze719 committed:
Refactor CI workflow to build in manylinux2014 container, consolidating installation and build steps into a single command for improved efficiency.

Katze719 committed:
Update CI workflow to include GCC and C++ compilers in manylinux2014 container for enhanced build capabilities.

Katze719 committed:
Update CI workflow to use GCC 11 and C++ compilers in manylinux2014 container for improved build compatibility.

Katze719 committed:
Refactor serial API to integrate cpp_core library, updating includes and error handling; remove legacy serial and status_codes headers, and adjust CMake configuration to link cpp_core for improved modularity.

Katze719 committed:
Update C API reference documentation to include detailed function descriptions, error handling examples, and new convenience functions for serial communication, enhancing clarity and usability for developers.

Mqxx commented on .github/workflows/ci.yml:
```diff
@@ -12,34 +12,24 @@ jobs:
         with:
-          name: libCPP-Unix-Bindings
-          path: build/libCPP-Unix-Bindings.so
+          name: libcpp_unix_bindings
+          path: build/libcpp_unix_bindings.so
```
```suggestion
          name: bindings_linux
          path: build/bindings_linux.so
```

Mqxx commented on .github/workflows/ci.yml:
```diff
@@ -12,34 +12,24 @@ jobs:
         uses: softprops/action-gh-release@v1
         with:
-          files: build/libCPP-Unix-Bindings.so 
+          files: build/libcpp_unix_bindings.so 
```
```suggestion
          files: build/bindings_linux.so 
```

Mqxx commented on CMakeLists.txt:
```diff
@@ -1,12 +1,19 @@
 set(VERSION_PATCH 0)
 
-set(PROJECT_N CPP-Unix-Bindings)
+set(PROJECT_N cpp_unix_bindings)
```
```suggestion
set(PROJECT_N cpp_bindings_linux)
```

Mqxx commented on README.md:
```diff
@@ -1,7 +1,7 @@
-# CPP-Unix-Bindings
+# cpp_unix_bindings
```
```suggestion
# C++ Bindings Linux
```

Mqxx commented on README.md:
```diff
@@ -1,7 +1,7 @@
 
 A compact C++23 library for talking to serial devices on Linux (e.g. Arduino).  
-The project builds a **shared library `libCPP-Unix-Bindings.so`** that can be used via
+The project builds a **shared library `libcpp_unix_bindings.so`** that can be used via
```
```suggestion
The project builds a **shared library `bindings_linux.so`** that can be used via
```

Mqxx commented on README.md:
```diff
@@ -21,7 +21,7 @@ cmake -S . -B build -DCMAKE_BUILD_TYPE=Release
 
 # The resulting library will be located at
-#   build/libCPP-Unix-Bindings.so
+#   build/libcpp_unix_bindings.so
```
```suggestion
#   build/bindings_linux.so
```

Mqxx commented on README.md:
````diff
@@ -36,7 +36,7 @@ Deno ships with a first-class FFI API.
 ```ts
 // serial_deno.ts
-const lib = Deno.dlopen('./build/libCPP-Unix-Bindings.so', {
+const lib = Deno.dlopen('./build/libcpp_unix_bindings.so', {
````
```suggestion
const lib = Deno.dlopen('./build/bindings_linux.so', {
```

Katze719 replied:
all libs under linux are prefixed with a `lib`, would not go against the default here

Katze719 replied:
go under `/usr/lib` or `/usr/lib64` in any linux distro and show me one lib that has not the prefix `lib` and is not built by you.

Katze719 committed:
Refactor serial API by removing legacy serial.cpp and implementing modular functions for serial communication, including read, write, and buffer management. Update CMake configuration to exclude serial.cpp and enhance project structure for improved maintainability.

Katze719 committed:
Refactor serialGetPortsInfo function to rename callback parameter from 'function' to 'callback_fn' for improved clarity and consistency in naming conventions.

Katze719 committed:
Add mutex protection to serial functions for thread safety

Introduce std::lock_guard to ensure thread-safe access to shared resources in serial communication functions, including buffer management and read/write operations. This change enhances the robustness of the API by preventing data races during concurrent access.

Katze719 committed:
Refactor CMake configuration and update serial API for modularity

- Simplified CMakeLists.txt by removing redundant versioning variables and consolidating project setup.
- Updated project to use C++23 standard and improved dependency management with CPM.
- Introduced new source file `serial_list_ports.cpp` for enhanced USB port information retrieval.
- Renamed `serialGetPortsInfo` to `serialListPorts` for clarity and consistency in the API.
- Adjusted unit tests to reflect the new function name and ensure proper functionality.

Katze719 committed:
Refactor serial API function signatures for consistency and readability

- Updated function signatures across multiple serial API functions to improve clarity by aligning parameter formatting and spacing.
- Enhanced error handling messages for better debugging and consistency in reporting invalid handles.
- Adjusted unit tests to reflect changes in function signatures and maintain code quality.

Katze719 committed:
Refactor serial API function signatures for improved consistency and readability

- Updated function signatures across multiple serial API functions to use 'auto' return type and improved parameter formatting.
- Replaced std::lock_guard with std::scoped_lock for better mutex management.
- Enhanced error handling messages for clarity and consistency in reporting invalid handles.

Mqxx commented on CMakeLists.txt:
```diff
@@ -1,11 +1,6 @@
-
-set(PROJECT_N cpp_unix_bindings)
-project(${PROJECT_N} VERSION ${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH})
+project(cpp_unix_bindings VERSION 1.0.0 LANGUAGES CXX)
```
```suggestion
project(
    cpp-bindings-linux
    VERSION 1.0.0
    DESCRIPTION "C++ Linux Bindings for the serial library"
    LANGUAGES CXX
)
```

Mqxx replied:
Also you should think about adding a `VERSION` variable.

Katze719 replied:
yes, got something cooking for that right here https://github.com/Katze719/cmake-git-versioning

Mqxx commented on CMakeLists.txt:
```diff
@@ -17,100 +12,61 @@ add_compile_options(-pthread)
-    OPTIONS "GTEST_BUILD_EXAMPLES OFF"
-    OPTIONS "GTEST_BUILD_DOCS OFF"
+    OPTIONS
+    "GTEST_BUILD_TESTS OFF"
+    "GTEST_BUILD_EXAMPLES OFF"
+    "GTEST_BUILD_DOCS OFF"
```
```suggestion
        "GTEST_BUILD_TESTS OFF"
        "GTEST_BUILD_EXAMPLES OFF"
        "GTEST_BUILD_DOCS OFF"
```

Mqxx commented on CMakeLists.txt:
```diff
@@ -17,100 +12,61 @@ add_compile_options(-pthread)
+target_link_libraries(${PROJECT_NAME} PRIVATE cpp_core)
+target_include_directories(${PROJECT_NAME} PUBLIC ${PROJECT_SOURCE_DIR}/src)
 
+set_target_properties(${PROJECT_NAME} PROPERTIES
+    VERSION ${PROJECT_VERSION}
+    SOVERSION ${PROJECT_VERSION_MAJOR}
+)
```
```suggestion
set_target_properties(
    ${PROJECT_NAME}
    PROPERTIES
        VERSION ${PROJECT_VERSION}
        SOVERSION ${PROJECT_VERSION_MAJOR}
)
```

Mqxx commented on CMakeLists.txt:
```diff
@@ -17,100 +12,61 @@ add_compile_options(-pthread)
+)
+
+# Test executable
 add_executable(tests
```
```suggestion
add_executable(
    tests
```

Mqxx commented on CMakeLists.txt:
```diff
@@ -17,100 +12,61 @@ add_compile_options(-pthread)
 target_include_directories(tests PRIVATE ${PROJECT_SOURCE_DIR}/src)
 
+# Copy shared library next to test binary
 add_custom_command(TARGET tests POST_BUILD
```
```suggestion
add_custom_command(
    TARGET tests POST_BUILD
```

Mqxx commented on CMakeLists.txt:
```diff
@@ -17,100 +12,61 @@ add_compile_options(-pthread)
-    COMMENT "Copy shared library next to aggregated test binary")
+    $<TARGET_FILE:${PROJECT_NAME}>
+    $<TARGET_FILE_DIR:tests>
+    COMMENT "Copy shared library next to test binary"
```
What is this comment?

Mqxx commented on src/serial_list_ports.cpp:
```diff
@@ -123,21 +123,20 @@ bool handleEntry(const fs::directory_entry& entry,
+                                                   const char* locationId,
+                                                   const char* productId,
+                                                   const char* vendorId),
+                               ErrorCallbackT error_callback)
```
Why is this formatting still so f*ked up?

```suggestion
extern "C" int serialListPorts(
    void (*callback_fn)(
        const char* port,
        const char* path,
        const char* manufacturer,
        const char* serialNumber,
        const char* pnpId,
        const char* locationId,
        const char* productId,
        const char* vendorId
    ),
    ErrorCallbackT error_callback
)
```

Katze719 replied:
this is technically just one long type

Katze719 replied:
will replace it with a using

Mqxx commented on docs/ffi_deno.md:
Vibe coded and generated as f*ck bro 😅 No one would use `…` as `...` in a description like that 😂

Mqxx commented on docs/ffi_deno.md:
Also you can remove this example, since documenting this is my job 👍 You can also remove the whole TypeScript `examples` folder.

Mqxx commented on examples/serial_advanced.ts:
Remove

Mqxx commented on examples/serial_callbacks.ts:
Remove

Mqxx commented on examples/serial_echo.ts:
Remove

Mqxx commented on src/serial_abort_read.cpp:
```diff
@@ -0,0 +1,26 @@
+
+using namespace serial_internal;
+
+extern "C" auto serialAbortRead(int64_t handle_ptr, ErrorCallbackT error_callback) -> int
```
Does it still make much sense to return an `int` if we have the error callback function? Also I don't see you use the `error_callback` anywhere.

Mqxx commented on src/serial_abort_read.cpp:
```diff
@@ -0,0 +1,26 @@
+#include <cstdint>
+#include <utility>
+
+using namespace serial_internal;
```
What is this for?

Mqxx commented on src/serial_in_bytes_total.cpp:
```diff
@@ -0,0 +1,21 @@
+
+using namespace serial_internal;
+
+extern "C" auto serialInBytesTotal(int64_t handle_ptr, ErrorCallbackT /*error_callback*/
+                                   ) -> int64_t
```
Formatting f*cked up.

Mqxx commented on src/serial_internal.hpp:
```diff
@@ -0,0 +1,133 @@
+// Wait until the file descriptor is ready for read/write (select based).
+// Returns >0 on ready, 0 on timeout, -1 on error.
+// ----------------------------------------------------------------------------
+inline int waitFdReady(int fd, int timeout_ms, bool want_write)
```
What is this function name? `waitFdReady`? Please use a proper name.

Mqxx commented on src/serial_internal.hpp:
Please try to refactor this file into its own `internal` directory.

Ex. `./internal/invoke_error.cpp`

Mqxx commented on src/serial_internal.hpp:
```diff
@@ -0,0 +1,133 @@
+// ----------------------------------------------------------------------------
+// Utility: map integer baud rate to POSIX speed_t.
+// ----------------------------------------------------------------------------
+inline speed_t to_speed_t(int baud)
```
What about custom baudrates? Currently we support custom baudrates. Or do we not?

Mqxx commented on src/serial_list_ports.cpp:
```diff
@@ -0,0 +1,172 @@
+
+} // namespace
+
+extern "C" auto serialListPorts(void (*callback_fn)(const char *port, const char *path, const char *manufacturer,
+                                                    const char *serial_number, const char *pnp_id,
+                                                    const char *location_id, const char *product_id,
+                                                    const char *vendor_id),
+                                ErrorCallbackT error_callback) -> int
```
Formatting f*cked up. Ti should be like this 👇

Mqxx commented on src/serial_list_ports.cpp:
```diff
@@ -0,0 +1,172 @@
+    const fs::path by_id_dir{"/dev/serial/by-id"};
+    if (!fs::exists(by_id_dir) || !fs::is_directory(by_id_dir))
+    {
+        invokeError(
+            std::to_underlying(cpp_core::StatusCodes::kNotFoundError),
+            "serialListPorts: Failed to get ports info",
+            error_callback
+        );
```
this

Mqxx commented on src/serial_open.cpp:
```diff
@@ -0,0 +1,133 @@
+
+using namespace serial_internal;
+
+extern "C" auto serialOpen(void *port, int baudrate, int data_bits, int parity, int stop_bits,
+                           ErrorCallbackT error_callback) -> intptr_t
```
Formatting f*cked up.

Mqxx commented on src/serial_out_bytes_total.cpp:
```diff
@@ -0,0 +1,21 @@
+
+using namespace serial_internal;
+
+extern "C" auto serialOutBytesTotal(int64_t handle_ptr, ErrorCallbackT /*error_callback*/
+                                    ) -> int64_t
```
Formatting f*cked up.

Mqxx commented on src/serial_read.cpp:
```diff
@@ -0,0 +1,61 @@
+
+using namespace serial_internal;
+
+extern "C" auto serialRead(int64_t handle_ptr, void *buffer, int buffer_size, int timeout_ms, int /*multiplier*/,
+                           ErrorCallbackT error_callback) -> int
```
Formatting f*cked up.

Mqxx commented on src/serial_read_until_sequence.cpp:
```diff
@@ -0,0 +1,69 @@
+
+using namespace serial_internal;
+
+extern "C" auto serialReadUntilSequence(int64_t handle_ptr, void *buffer, int buffer_size, int timeout_ms,
+                                        int multiplier, void *sequence_ptr, ErrorCallbackT error_callback) -> int
```
Formatting f*cked up.

Mqxx commented on src/serial_set_error_callback.cpp:
```diff
@@ -0,0 +1,9 @@
+
+#include <cpp_core/interface/serial_set_error_callback.h>
+
+extern "C" void serialSetErrorCallback(ErrorCallbackT error_callback)
```
Not sure if this function is still usefull, if we have an error callback function that we pass to each function individually.

Mqxx commented on src/serial_set_read_callback.cpp:
```diff
@@ -0,0 +1,8 @@
+
+#include <cpp_core/interface/serial_set_read_callback.h>
+
+extern "C" void serialSetReadCallback(void (*callback_fn)(int))
```
Same here. I am not sure if this function is usefull at all.

Mqxx commented on src/serial_set_write_callback.cpp:
```diff
@@ -0,0 +1,8 @@
+
+#include <cpp_core/interface/serial_set_write_callback.h>
+
+extern "C" void serialSetWriteCallback(void (*callback_fn)(int))
```
Same here. I am not sure if this function is usefull at all.

Mqxx commented on src/serial_state.cpp:
Why is this its own file, but the rest is crammed into one internal file (`serial_internal.hpp`) above?

Mqxx commented on tests/serial_unit_tests.cpp:
Please refactor to individual unit test files next to the actual files.

Ex. File to test: `serial_read.cpp`, Unit test file next to it: `serial_read.test.cpp`

I know, for archiving this we(you) need to change the CMake again, but please do it that way. Adding/Changing/Removing indivitual tests that way is much easier and more clearer in terms of what you are doing/testing.
